src/librustc_passes/reachable.rs

```diff
@@ -73,26 +73,28 @@ struct ReachableContext<'a, 'tcx> {
     any_library: bool,
 }
 
-impl<'a, 'tcx> Visitor<'tcx> for ReachableContext<'a, 'tcx> {
+impl<'tcx> Visitor<'tcx> for ReachableContext<'tcx> {
     type Map = intravisit::ErasedMap<'tcx>;
 
     fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
         NestedVisitorMap::None
     }
 
     fn visit_nested_body(&mut self, body: hir::BodyId) {
-        let old_tables = self.tables;
-        self.tables = self.tcx.body_tables(body);
+        let old_tables = self.tables.replace(self.tcx.body_tables(body));
         let body = self.tcx.hir().body(body);
         self.visit_body(body);
         self.tables = old_tables;
     }
 
     fn visit_expr(&mut self, expr: &'tcx hir::Expr<'tcx>) {
         let res = match expr.kind {
-            hir::ExprKind::Path(ref qpath) => Some(self.tables.qpath_res(qpath, expr.hir_id)),
+            hir::ExprKind::Path(ref qpath) => {
+                Some(self.tables.unwrap().qpath_res(qpath, expr.hir_id))
+            }
             hir::ExprKind::MethodCall(..) => self
                 .tables
+                .unwrap()
```

Comment:
As noted above, `hir::Expr` implies we're in a `hir::Body`. The method isn't a bad idea, and it's how I've done it for lints (via #73743).

But there's a few problems with that:
* there's half a dozen types that would need the same method, do you want me to just copy-paste it?
* for `rustc_lint::LateContext` at least there's no field named `tables`, but for the rest I would have to rename `tables` fields to `maybe_tables`, perhaps? (that would fit with the `fn maybe_tables` I added to `LateContext`) - note that this is just about forcing the method to be used, and making it clearer that it should be used
* the comment could list out the situations in which it can be safely used, but you could still accidentally call it outside of a body (`visit_qpath`/`visit_path` being the prime candidate for mistakes here)

I guess they all have solutions, so it doesn't seem to bad, just wanted to confirm all that.

Reply:
The point of the method, in my mind, is mostly to house the comment in a logical, central place, not so much to avoid bugs. I'd probably include it but I'm ok either way.

Reply:
Maybe a name like `tables_from_within_body` or something *might* be suggestive as to the reason it's safe to call, but then also kind of "ponderous".